File: lib/iriq/explanation.rb
module Iriq
  # Builds a per-segment explanation for a single identifier.
  #
  #   Explanation.explain("https://foo.com/users/123")
  #   # => [
  #   #      { value: "users", type: :literal,    variable: false, hint: nil       },
  #   #      { value: "123",   type: :integer_id, variable: true,  hint: "user_id" },
  #   #    ]
  module Explanation
    module_function

    def explain(input, classifier: SegmentClassifier::DEFAULT)
      iri = input.is_a?(Identifier) ? input : Parser.parse(input)

      if iri.urn?
        explain_urn(iri, classifier)
      else
        SegmentHints.derive(iri.path_segments, classifier)
      end
    end

    def explain_urn(iri, classifier)
      return [] unless iri.nss

      parts = iri.nss.include?(":") ? iri.nss.split(":", 2) : [iri.nss]
      SegmentHints.derive(parts, classifier)
    end
  end
end

File: rust/iriq/src/explanation.rs
use crate::classifier::{SegmentClassifier, DEFAULT_CLASSIFIER};
use crate::errors::ParseError;
use crate::hints::{derive_hints, SegmentHint};
use crate::identifier::Identifier;
use crate::parser::parse;

pub fn explain(input: &str) -> Result<Vec<SegmentHint>, ParseError> {
    let iri = parse(input)?;
    Ok(explain_identifier(&iri, &DEFAULT_CLASSIFIER))
}

pub fn explain_identifier(iri: &Identifier, c: &SegmentClassifier) -> Vec<SegmentHint> {
    if iri.is_urn() {
        return explain_urn(iri, c);
    }
    derive_hints(&iri.path_segments, c)
}

fn explain_urn(iri: &Identifier, c: &SegmentClassifier) -> Vec<SegmentHint> {
    if iri.nss.is_empty() {
        return Vec::new();
    }
    let parts: Vec<String> = if let Some((ns, val)) = iri.nss.split_once(':') {
        vec![ns.to_string(), val.to_string()]
    } else {
        vec![iri.nss.clone()]
    };
    derive_hints(&parts, c)
}
